frontend/src/components/TradeBox/EncryptedChat/EncryptedTrutleChat/index.tsx

```diff
@@ -0,0 +1,303 @@
+import React, { useEffect, useState } from 'react';
+import { useTranslation } from 'react-i18next';
+import { Button, Tooltip, TextField, Grid, Container, Paper, Typography } from '@mui/material';
+import { encryptMessage, decryptMessage } from '../../../../pgp';
+import { saveAsJson } from '../../../../utils';
+import { AuditPGPDialog } from '../../../Dialogs';
+import { systemClient } from '../../../../services/System';
+import { websocketClient, WebsocketConnection } from '../../../../services/Websocket';
+
+// Icons
+import CircularProgress from '@mui/material/CircularProgress';
+import KeyIcon from '@mui/icons-material/Key';
+import { ExportIcon } from '../../../Icons';
+import { useTheme } from '@mui/system';
+import MessageCard from '../MessageCard';
+import ChatHeader from '../ChatHeader';
+import { EncryptedChatMessage, ServerMessage } from '..';
+import { apiClient } from '../../../../services/api';
+import ChatBottom from '../ChatBottom';
+
+interface Props {
+  orderId: number;
+  userNick: string;
+  takerNick: string;
+  chatOffset: number;
+  messages: EncryptedChatMessage[];
+  setMessages: (messages: EncryptedChatMessage[]) => void;
+}
+
+const EncryptedTurtleChat: React.FC<Props> = ({
+  orderId,
+  userNick,
+  takerNick,
+  chatOffset,
+  messages,
+  setMessages,
+}: Props): JSX.Element => {
+  const { t } = useTranslation();
+  const theme = useTheme();
+
+  const audio = new Audio(`/static/assets/sounds/chat-open.mp3`);
+  const [peerConnected, setPeerConnected] = useState<boolean>(false);
+  const [ownPubKey] = useState<string>(
+    (systemClient.getCookie('pub_key') ?? '').split('\\').join('\n'),
+  );
+  const [ownEncPrivKey] = useState<string>(
+    (systemClient.getCookie('enc_priv_key') ?? '').split('\\').join('\n'),
+  );
+  const [peerPubKey, setPeerPubKey] = useState<string>();
+  const [token] = useState<string>(systemClient.getCookie('robot_token') || '');
+  const [value, setValue] = useState<string>('');
+  const [audit, setAudit] = useState<boolean>(false);
+  const [waitingEcho, setWaitingEcho] = useState<boolean>(false);
+  const [lastSent, setLastSent] = useState<string>('---BLANK---');
+  const [messageCount, setMessageCount] = useState<number>(0);
+  const [serverMessages, setServerMessages] = useState<ServerMessage[]>([]);
+  const [lastIndex, setLastIndex] = useState<number>(0);
+
+  useEffect(() => {
+    if (messages.length > messageCount) {
+      audio.play();
+      setMessageCount(messages.length);
+    }
+  }, [messages, messageCount]);
+
+  useEffect(() => {
+    if (serverMessages.length > 0 && peerPubKey) {
+      serverMessages.forEach(onMessage);
+    }
+  }, [serverMessages, peerPubKey]);
+
+  useEffect(() => {
+    if (chatOffset > lastIndex) {
+      loadMessages();
+    }
+  }, [chatOffset]);
+
+  const loadMessages: () => void = () => {
+    apiClient.get(`/api/chat?order_id=${orderId}&offset=${lastIndex}`).then((results: any) => {
```

Comment:
This call is not needed. The `chat` object already is in `this.props.data.chat` (it is returned with the `/api/order` call when the offset is supplied).

Reply:
It's hard to maintain the data coming from the parent and a good alignment between both turtle and websocket mode components, I did at first but didn't like at all. What if we keep this and just return a number instead of array for the `/order` endpoint?

Reply:
I think it is an extra level of difficulty to have this API call if it is not needed, right? It also makes more API requests than necessary and increases latency. I might not be seeing the full picture.